File: 2976-minimum-cost-to-convert-string-i/2976-minimum-cost-to-convert-string-i.java
class Solution {
    public long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {
        int[][]matrix=new int[26][26];
        for(int i=0;i<26;i++){
            Arrays.fill(matrix[i],(int)(1e9));
            matrix[i][i]=0;
        }
        for(int i=0;i<cost.length;i++){
            matrix[original[i]-'a'][changed[i]-'a']=Math.min(cost[i],matrix[original[i]-'a'][changed[i]-'a']);
        }
        for(int k=0;k<26;k++){
            for(int i=0;i<26;i++){
                for(int j=0;j<26;j++){
                    if(matrix[i][k]==1e9 || matrix[k][j]==1e9)continue;
                    matrix[i][j]=Math.min(matrix[i][j],matrix[i][k]+matrix[k][j]);
                }
            }
        }
        long ans=0;
        for(int i=0;i<source.length();i++){
            if(matrix[source.charAt(i)-'a'][target.charAt(i)-'a']==1e9){
                return (long)(-1);
            }else{
                ans+=matrix[source.charAt(i)-'a'][target.charAt(i)-'a'];
            }
        }
        return ans;
    }
}

File: 2976-minimum-cost-to-convert-string-i/2976-minimum-cost-to-convert-string-i.js
/**
 * @param {string} source
 * @param {string} target
 * @param {character[]} original
 * @param {character[]} changed
 * @param {number[]} cost
 * @return {number}
 */
var minimumCost = function(source, target, original, changed, cost) {
    const mat = Array.from({ length: 26 }, () => Array(26).fill(Infinity));
        
    for (let i = 0; i < cost.length; i++) {
        const s = original[i].charCodeAt(0) - 'a'.charCodeAt(0);
        const e = changed[i].charCodeAt(0) - 'a'.charCodeAt(0);
        mat[s][e] = Math.min(mat[s][e], cost[i]);
    }
    
    for (let i = 0; i < 26; i++) {
        mat[i][i] = 0;
    }

    for (let k = 0; k < 26; k++) {
        for (let i = 0; i < 26; i++) {
            if (mat[i][k] < Infinity) {
                for (let j = 0; j < 26; j++) {
                    if (mat[k][j] < Infinity) {
                        mat[i][j] = Math.min(mat[i][j], mat[i][k] + mat[k][j]);
                    }
                }
            }
        }
    }
    
    let total = 0;
    for (let i = 0; i < source.length; i++) {
        const s = source[i].charCodeAt(0) - 'a'.charCodeAt(0);
        const e = target[i].charCodeAt(0) - 'a'.charCodeAt(0);
        if (s !== e) {
            if (mat[s][e] === Infinity) return -1;
            total += mat[s][e];
        }
    }
    return total;
};